src/clients/arclib/arccat.1.in
.\" -*- nroff -*-
.TH ARCCAT 1 "@DATE@" "NorduGrid ARC @VERSION@" "NorduGrid Users Manual"

.SH NAME

arccat \- ARC Catenate

.SH DESCRIPTION

The
.B arccat
command can be used to view the stdout or stderr of a running job. It
can also be used to show A-REX's error log of a job.

.SH SYNOPSIS

.B arccat [options] [job ...]

.SH OPTIONS

.IP "\fB-a\fR, \fB--all\fR"
all jobs
.IP "\fB-j\fR, \fB--joblist\fR=\fIfilename\fR"
the file storing information about active jobs (default ~/.arc/jobs.xml)
.IP "\fB-i\fR, \fB--jobids-from-file\fR=\fIfilename\fR"
a file containing a list of jobIDs
.IP "\fB-c\fR, \fB--cluster\fR=\fIname\fR"
select one or more computing elements: \fIname\fR can be an alias for a single CE, a group of CEs or a URL
.IP "\fB-I\fR, \fB--infointerface\fR=\fIInterfaceName\fR"
the computing element specified by URL at the command line should be
queried using this information interface (possible options:
org.nordugrid.ldapng, org.nordugrid.ldapglue2,
org.nordugrid.wsrfglue2, org.ogf.glue.emies.resourceinfo)
.IP "\fB-r\fR, \fB--rejectmanagement\fR=\fIURL\fR"
skip jobs which are on a computing element with a given URL
.IP "\fB-s\fR, \fB--status\fR=\fIstatusstr\fR"
only select jobs whose status is statusstr
.IP "\fB-o\fR, \fB--stdout\fR"
show the stdout of the job (default)
.IP "\fB-e\fR, \fB--stderr\fR"
show the stderr of the job
.IP "\fB-l\fR, \fB--joblog\fR"
show A-REX's error log of the job
.IP "\fB-P\fR, \fB--listplugins\fR"
list the available plugins
.IP "\fB-t\fR, \fB--timeout\fR=\fIseconds\fR"
timeout in seconds (default 20)
.IP "\fB-z\fR, \fB--conffile\fR=\fIfilename\fR"
configuration file (default ~/.arc/client.conf)
.IP "\fB-d\fR, \fB--debug\fR=\fIdebuglevel\fR"
FATAL, ERROR, WARNING, INFO, VERBOSE or DEBUG
.IP "\fB-v\fR, \fB--version\fR"
print version information
.IP "\fB-?\fR, \fB--help\fR"
print help
.LP

.SH ARGUMENTS

.IP "\fBjob\fR ..."
list of jobids and/or jobnames
.LP

.SH EXTENDED DESCRIPTION

The
.B arccat
command displays the stdout or stderr of running jobs. It can also
display A-REX's error log of a job. The job can be referred to either
by the jobid that was returned by
.BR arcsub (1)
at submission time or by its jobname if the job description that was
submitted contained a jobname attribute.

More than one jobid and/or jobname can be given. If several jobs were
submitted with the same jobname the stdout, stderr or A-REX error log
of all those jobs are shown. If the
.B --joblist
option is used the list of jobs is read from a file with the
specified filename. By specifying the
.B --all
option, the stdout, stderr or A-REX error log of all active jobs will
be shown.

The
.B --cluster
option can be used to select or reject jobs at specific clusters. See
.BR arcsub (1)
for a discussion of the format of arguments to this option. The
.B --status
option can be used to select jobs in a specific state. These options
can be repeated several times.

Only jobs where the stdout or stderr argument was given in the job
description can display the contents of those files.

.SH FILES

.TP
.B ~/.arc/client.conf
Some options can be given default values by specifying them in the
ARC client configuration file. By using the
.B --conffile
option a different configuration file can be used than the default.

.TP
.B ~/.arc/jobs.xml
This a local list of the user's active jobs. When a job is
successfully submitted it is added to this list and when it is removed
from the remote cluster it is removed from this list. This list is
used as the list of all active jobs when the user specifies the
.B --all
option to the various NorduGrid ARC user interface commands. By using the
.B --joblist
option a different file can be used than the default.

.SH ENVIRONMENT VARIABLES

.TP
.B X509_USER_PROXY
The location of the user's grid proxy file. Shouldn't be set unless
the proxy is in a non-standard location.

.TP
.B ARC_LOCATION
The location where ARC is installed can be specified by this variable. If not
specified the install location will be determined from the path to the
command being executed, and if this fails a WARNING will be given stating the
location which will be used.

.TP
.B ARC_PLUGIN_PATH
The location of ARC plugins can be specified by this variable. Multiple
locations can be specified by separating them by : (; in Windows). The
default location is \fB$ARC_LOCATION\fR/lib/arc (\\ in Windows).

.SH COPYRIGHT
APACHE LICENSE Version 2.0

.SH AUTHOR

Bjarte Mohn <bjarte.mohn@fysast.uu.se>

.SH SEE ALSO

.BR arcclean (1),
.BR arccp (1),
.BR arcget (1),
.BR arcinfo (1),
.BR arckill (1),
.BR arcls (1),
.BR arcmigrate (1),
.BR arcmkdir (1),
.BR arcproxy (1),
.BR arcrenew (1),
.BR arcresub (1),
.BR arcresume (1),
.BR arcrm (1),
.BR arcstat (1),
.BR arcsub (1),
.BR arcsync (1),
.BR arctest (1)



NEWS
no news good news :)


src/clients/compute/arccat.cpp
// -*- indent-tabs-mode: nil -*-

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <iostream>
#include <list>
#include <string>

#include <arc/ArcLocation.h>
#include <arc/IString.h>
#include <arc/Logger.h>
#include <arc/StringConv.h>
#include <arc/compute/JobControllerPlugin.h>
#include <arc/compute/JobSupervisor.h>
#include <arc/UserConfig.h>

#include "utils.h"

int RUNMAIN(arccat)(int argc, char **argv) {

  setlocale(LC_ALL, "");

  Arc::Logger logger(Arc::Logger::getRootLogger(), "arccat");
  Arc::LogStream logcerr(std::cerr);
  logcerr.setFormat(Arc::ShortFormat);
  Arc::Logger::getRootLogger().addDestination(logcerr);
  Arc::Logger::getRootLogger().setThreshold(Arc::WARNING);

  Arc::ArcLocation::Init(argv[0]);

  ClientOptions opt(ClientOptions::CO_CAT,
                    istring("[job ...]"),
                    istring("The arccat command performs the cat "
                            "command on the stdout, stderr or grid\n"
                            "manager's error log of the job."));

  std::list<std::string> jobidentifiers = opt.Parse(argc, argv);

  if (opt.showversion) {
    std::cout << Arc::IString("%s version %s", "arccat", VERSION)
              << std::endl;
    return 0;
  }

  // If debug is specified as argument, it should be set before loading the configuration.
  if (!opt.debug.empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(opt.debug));

  if (opt.show_plugins) {
    std::list<std::string> types;
    types.push_back("HED:JobControllerPlugin");
    showplugins("arccat", types, logger);
    return 0;
  }

  Arc::UserConfig usercfg(opt.conffile, opt.joblist);
  if (!usercfg) {
    logger.msg(Arc::ERROR, "Failed configuration initialization");
    return 1;
  }

  if (!checkproxy(usercfg)) {
    return 1;
  }

  if (opt.debug.empty() && !usercfg.Verbosity().empty())
    Arc::Logger::getRootLogger().setThreshold(Arc::string_to_level(usercfg.Verbosity()));

  for (std::list<std::string>::const_iterator it = opt.jobidinfiles.begin(); it != opt.jobidinfiles.end(); it++) {
    if (!Arc::Job::ReadJobIDsFromFile(*it, jobidentifiers)) {
      logger.msg(Arc::WARNING, "Cannot read specified jobid file: %s", *it);
    }
  }

  if (opt.timeout > 0)
    usercfg.Timeout(opt.timeout);

  if ((!opt.joblist.empty() || !opt.status.empty()) && jobidentifiers.empty() && opt.clusters.empty())
    opt.all = true;

  if (jobidentifiers.empty() && opt.clusters.empty() && !opt.all) {
    logger.msg(Arc::ERROR, "No jobs given");
    return 1;
  }

  std::list<std::string> selectedURLs;
  if (!opt.clusters.empty()) {
    selectedURLs = getSelectedURLsFromUserConfigAndCommandLine(usercfg, opt.clusters);
  }
  std::list<std::string> rejectManagementURLs = getRejectManagementURLsFromUserConfigAndCommandLine(usercfg, opt.rejectmanagement);

  std::list<Arc::Job> jobs;
  if (!Arc::Job::ReadJobsFromFile(usercfg.JobListFile(), jobs, jobidentifiers, opt.all, selectedURLs, rejectManagementURLs)) {
    logger.msg(Arc::ERROR, "Unable to read job information from file (%s)", usercfg.JobListFile());
    return 1;
  }

  for (std::list<std::string>::const_iterator itJIdentifier = jobidentifiers.begin();
       itJIdentifier != jobidentifiers.end(); ++itJIdentifier) {
    std::cout << Arc::IString("Warning: Job not found in job list: %s", *itJIdentifier) << std::endl;
  }

  Arc::JobSupervisor jobmaster(usercfg, jobs);
  jobmaster.Update();
  jobmaster.SelectValid();
  if (!opt.status.empty()) {
    jobmaster.SelectByStatus(opt.status);
  }

  jobs = jobmaster.GetSelectedJobs();
  if (jobs.empty()) {
    std::cout << Arc::IString("No jobs") << std::endl;
    return 1;
  }

  std::string resourceName;
  Arc::Job::ResourceType resource;
  if (opt.show_joblog)      { resource = Arc::Job::JOBLOG; resourceName = "joblog"; }
  else if (opt.show_stderr) { resource = Arc::Job::STDERR; resourceName = "stderr"; }
  else                      { resource = Arc::Job::STDOUT; resourceName = "stdout"; }

  // saving to a temp file is necessary because chunks from server
  // may arrive out of order
  std::string filename = Glib::build_filename(Glib::get_tmp_dir(), "arccat.XXXXXX");
  int tmp_h = Glib::mkstemp(filename);
  if (tmp_h == -1) {
    logger.msg(Arc::INFO, "Could not create temporary file \"%s\"", filename);
    logger.msg(Arc::ERROR, "Cannot create output of %s for any jobs", resourceName);
    return 1;
  }

  Arc::URL dst("stdio:///"+Arc::tostring(tmp_h));
  if (!dst) {
    logger.msg(Arc::ERROR, "Cannot create output of %s for any jobs", resourceName);
    logger.msg(Arc::INFO, "Invalid destination URL %s", dst.str());
    close(tmp_h);
    unlink(filename.c_str());
    return 1;
  }

  int retval = 0;
  for (std::list<Arc::Job>::const_iterator it = jobs.begin();
       it != jobs.end(); it++) {
    if (!it->State || (!opt.status.empty() &&
        std::find(opt.status.begin(), opt.status.end(), it->State()) == opt.status.end() &&
        std::find(opt.status.begin(), opt.status.end(), it->State.GetGeneralState()) == opt.status.end())) {
      continue;
    }

    if (it->State == Arc::JobState::DELETED) {
      logger.msg(Arc::WARNING, "Job deleted: %s", it->JobID);
      retval = 1;
      continue;
    }

    // The job-log might be available before the job has started (middleware dependent).
    if (!opt.show_joblog &&
        !it->State.IsFinished() &&
        it->State != Arc::JobState::RUNNING &&
        it->State != Arc::JobState::FINISHING) {
      logger.msg(Arc::WARNING, "Job has not started yet: %s", it->JobID);
      retval = 1;
      continue;
    }

    if ((opt.show_joblog && it->LogDir.empty()) ||
        (!opt.show_joblog && opt.show_stderr && it->StdErr.empty()) ||
        (!opt.show_joblog && !opt.show_stderr && it->StdOut.empty())) {
      logger.msg(Arc::ERROR, "Cannot determine the %s location: %s", resourceName, it->JobID);
      retval = 1;
      continue;
    }

    Arc::URL src;
    it->GetURLToResource(resource, src);
    if (!src) {
      logger.msg(Arc::ERROR, "Cannot create output of %s for job (%s): Invalid source %s", resourceName, it->JobID, src.str());
      retval = 1;
      continue;
    }

    if (!it->CopyJobFile(usercfg, src, dst)) {
      retval = 1;
      continue;
    }

    logger.msg(Arc::VERBOSE, "Catting %s for job %s", resourceName, it->JobID);

    std::ifstream is(filename.c_str());
    char c;
    while (is.get(c)) {
      std::cout.put(c);
    }
    is.close();
  }

  close(tmp_h);
  unlink(filename.c_str());

  return retval;
}
